contents/blog/clickhouse-materialized-columns.md

```diff
@@ -0,0 +1,126 @@
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+On a large test dataset this query takes a while complete, while without the url filter the query is almost instant. Adding even more filters just makes the query slower and slower. Let’s dig in why!
+
+## Looking at flamegraphs
+
+ClickHouse has great tools for introspecting queries. Looking at `system.query_log`  we can see that the query:
+
+- Took 3433ms
+- Read 79.17 GiB from disk
+
+To dig even deeper, we can use clickhouse-flamegraph to peek into what the CPU is doing during query execution.
+
+[![Flamegraph](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)](../images/blog/clickhouse-materialized-columns/query-json-extract-CPU.svg)
+
+From this we can see that the clickhouse node CPU is spending most of its time parsing JSON.
+
+The typical solution would be to extract $current_url to a separate column. This would get rid of the JSON parsing and reduce the amount of data read from disk.
+
+However, in this particular case it wouldn’t work because:
+
+1. The data is passed from users - meaning we’d end up with millions (!) of unique columns
+2. This would complicate live data ingestion a lot, introducing new and exciting race conditions
+
+
+## Enter materialized columns
+
+A less known feature of ClickHouse is the concept of materialized columns. You can add one as follows:
+
+```sql
+ALTER TABLE events
+ADD COLUMN mat_$current_url
+VARCHAR MATERIALIZED JSONExtractString(properties_json, '$current_url')
+```
+
+This will create a new column that will be automatically filled for incoming data. Under the hood ClickHouse stores the
+
+Running the query again would immediately not yet speed up the query since historical data is unaffected. The way around this is running an [OPTIMIZE](https://clickhouse.tech/docs/en/sql-reference/statements/optimize/) command [1] :
+
+```sql
+OPTIMIZE TABLE events FINAL
+```
+
+Running the updated query speeds things up significantly:
+
+```sql
+SELECT count(*)
+FROM events
+WHERE event = '$pageview'
+  AND mat_$current_url = 'https://app.posthog.com/login'
+  AND timestamp >= '2021-08-01'
+  AND timestamp < '2021-09-01'
+```
+
+Looking at `system.query_log`, the new query:
+
+- took 980ms (**71% improvement**)
+- read 14.36 GiB from disk (**81% improvement**) [2]
+
+The wins are even more magnified if more than one property filter is used at a time.
+
+
+## Usage at PostHog
+
+PostHog as an analytics tool allows users to slice and dice their data in many ways across huge time ranges and datasets. This also means that performance when investigating things is of key importance but also that we currently do nearly no preaggregation.
+
+Rather than materialize all columns, built a solution which looks at recent slow queries using `system.query_log`, determines which properties need materializing from there and backfills the data on a weekend. [3] [4]
```

Comment:
We're open source - so code links everywhere would be cool

Reply:
Done!